pkg/restmapper/lazyrestmapper.go

```diff
@@ -10,40 +10,124 @@ import (
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/restmapper"
-	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 )
 
-type RESTMapperProvider func(c *rest.Config) (meta.RESTMapper, error)
-
-// NewPartialRestMapperProvider returns configured 'partial' rest mapper provider intended to be used with controller-runtime manager.
-// Takes GroupFilterPredicate as an argument for filtering out APIGroups during discovery procedure.
-func NewPartialRestMapperProvider(groupFilterPredicate GroupFilterPredicate) RESTMapperProvider {
-	partialRESTMapperProvider := func(c *rest.Config) (meta.RESTMapper, error) {
-		drm, err := apiutil.NewDynamicRESTMapper(c,
-			apiutil.WithLazyDiscovery,
-			apiutil.WithCustomMapper(
-				func() (meta.RESTMapper, error) {
-					dc, err := discovery.NewDiscoveryClientForConfig(c)
-					if err != nil {
-						return nil, err
-					}
-
-					groupResources, err := getFilteredAPIGroupResources(dc, groupFilterPredicate)
-					if err != nil {
-						return nil, err
-					}
-
-					return restmapper.NewDiscoveryRESTMapper(groupResources), nil
-				},
-			),
-		)
-		if err != nil {
-			return nil, err
+// LazyRESTMapper is a RESTMapper that will lazily query the provided
+// client for discovery information to do REST mappings.
+type LazyRESTMapper struct {
+	mapper      meta.RESTMapper
+	client      *discovery.DiscoveryClient
+	knownGroups map[string]any
+
+	// mutex to provide thread-safe mapper reloading
+	mu sync.Mutex
+}
+
+// NewLazyRESTMapper initializes a LazyRESTMapper
+func NewLazyRESTMapper(c *rest.Config) (meta.RESTMapper, error) {
+	dc, err := discovery.NewDiscoveryClientForConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return &LazyRESTMapper{
+		mapper:      restmapper.NewDiscoveryRESTMapper([]*restmapper.APIGroupResources{}),
+		client:      dc,
+		knownGroups: map[string]any{},
+	}, nil
+}
+
+func (m *LazyRESTMapper) addKnownGroupAndReload(groupName string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.knownGroups[groupName] = struct{}{}
+
+	groupResources, err := getFilteredAPIGroupResources(m.client, m.knownGroups)
```

Comment:
Hm, is it possible to get `[]*restmapper.APIGroupResources` from the existing mapper, append the `APIGroupResources` for the single group there, and reinitialize the mapper? This might allow not to make requests for all groups in the `knownGroups` and make it just for a new one.

Or, alternatively, `[]*restmapper.APIGroupResources` which comes from `getFilteredAPIGroupResources` might be cached in the `LazyRESTMapper` somehow (raw idea: perhaps we could use `knownGroups` map for this, like `map[string]*restmapper.APIGroupResources`).